File: gui/iode_objs/edit/edit_table_delegate.cpp
#pragma once
#include "edit_table_delegate.h"


bool EditTableDelegate::is_editable(const QModelIndex& index) const
{
	const EditTableModel* table_model = static_cast<const EditTableModel*>(index.model());
	TableLineType line_type = table_model->get_line_type(index.row());
	return (line_type == TABLE_LINE_TITLE) || (line_type == TABLE_LINE_CELL);
}

QWidget* EditTableDelegate::createEditor(QWidget* parent, const QStyleOptionViewItem& option, const QModelIndex& index) const
{
	if (is_editable(index))
	{
		QLineEdit* editor = new QLineEdit(parent);
		return editor;
	}
	else
	{
		return nullptr;
	}
}

void EditTableDelegate::setEditorData(QWidget* editor, const QModelIndex& index) const
{
	if (is_editable(index))
	{
		QString value = index.model()->data(index, Qt::EditRole).toString();
		QLineEdit* lineEdit = static_cast<QLineEdit*>(editor);
		lineEdit->setText(value);
	}
}

void EditTableDelegate::setModelData(QWidget* editor, QAbstractItemModel* model, const QModelIndex& index) const
{
	if (is_editable(index))
	{
		QLineEdit* lineEdit = static_cast<QLineEdit*>(editor);
		QString value = lineEdit->text();
		model->setData(index, value, Qt::EditRole);
	}
}

File: gui/iode_gui/iode_objs/edit/edit_table_delegate.py
from PySide6.QtCore import Qt, QModelIndex, QAbstractItemModel
from PySide6.QtWidgets import QStyledItemDelegate, QWidget, QStyleOptionViewItem, QLineEdit
from .edit_table_model import EditTableModel

from iode import TableLineType

class EditTableDelegate(QStyledItemDelegate):

    def __init__(self, parent=None):
        super().__init__(parent)

    def is_editable(self, index: QModelIndex):
        table_model: EditTableModel = index.model()
        line_type = table_model.line_type(index.row())
        return line_type == TableLineType.TITLE or line_type == TableLineType.CELL

    def createEditor(self, parent: QWidget, option: QStyleOptionViewItem, index: QModelIndex):
        if self.is_editable(index):
            editor = QLineEdit(parent)
            return editor
        else:
            return None

    def setEditorData(self, editor: QWidget, index: QModelIndex):
        if self.is_editable(index):
            value = index.model().data(index, Qt.ItemDataRole.EditRole)
            lineEdit: QLineEdit = editor
            lineEdit.setText(value)

    def setModelData(self, editor: QWidget, model: QAbstractItemModel, index: QModelIndex):
        if self.is_editable(index):
            lineEdit: QLineEdit = editor
            value = lineEdit.text()
            model.setData(index, value, Qt.ItemDataRole.EditRole)
